### PR TITLE
Support of the Nezha Allwinner D1 - Yocto kirkstone

### PR DESCRIPTION
**Pull request by Cezarus27**
This PR provides early support to the Nezha Allwinner D1 board. It uses **dev** versions of the Linux kernel, U-Boot and OpenSBI. This version was checked and it builds with success.

Key changes:
**- nezha.yml: add file used with kas-docker**

**- linux-nezha-dev | u-boot-nezha | boot0: add patch which fix build with binutils 2.28**

**- nezha.wks: description of SD card image for Nezha D1 development board**

**- linux-nezha-dev: use custom version of kernel with paches for D1 chip**

**- u-boot-nezha: add patch which fix build with binutils 2.28**

**- u-boot-nezha: add patch which increases the CONFIF_SYS_BOOTM_LEN**

**- toc.cfg: add configuration file of TOC1 U-Boot image**

**- uEnv-nezha.txt: U-Boot bootargs for Nezha board**

**- u-boot-nezha.bb: add recipe with patches for Nezha board**

**- opensbi: update mainline with patches to fit Nezha board**

**- boot0: add patch which fix build with binutils 2.28**

**- boot0: add patch for Makefile to fit it with yocto build environment**

**- boot0.bb: add recipe of the Nezha SPL**

**- nezha-allwinner-d1.conf: add machine configuration for Nezha board**
 
External resources:
- [linux-sunxi](https://linux-sunxi.org/Allwinner_Nezha)
- [meta-nezha](https://github.com/Cezarus27/meta-nezha)
- [Nezha SBC first impression](https://blog.3mdeb.com/2021/2021-11-19-nezha-riscv-sbc-first-impression/)

Known issues:
- no support of the WiFi&BT driver `XR829` in the currently used kernel
- `rng-tools` not working. For some reason, it crushes during the start with `SIGSEGV` in `libc`. This problem doesn't exist when the `haveged` random number generator is used in the build.

**Signed-off-by: Cezary Sobczak <cezary.sobczak@3mdeb.com>**

### CODE REVIEW

File: recipes-bsp/boot0/boot0.bb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+LIC_FILES_CHKSUM = "file://${COREBASE}/meta/files/common-licenses/GPL-2.0-or-later;md5=fed54355545ffd980b814dab4a3b312c"
+
+FILESEXTRAPATHS:prepend := "${THISDIR}/files:"
+SRC_URI = " \
+	git://github.com/smaeul/sun20i_d1_spl;protocol=https;branch=mainline \
+	file://0001-config.mk-provide-path-to-linux-headers.patch \
+	file://0002-riscv-fix-build-with-binutils-2.38.patch \
+"
+
+PV = "1.0-dev"
```
Comment (kraj) [inline recipes-bsp/boot0/boot0.bb:12]: better to use `PV = "1.0+git${SRCPV}"` here

File: recipes-bsp/boot0/boot0.bb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+SRCREV = "0ad88bfdb723b1ac74cca96122918f885a4781ac"
+
+S = "${WORKDIR}/git"
+
+CFLAGS += " --sysroot=${RECIPE_SYSROOT}"
```
Comment (kraj) [inline recipes-bsp/boot0/boot0.bb:17]: --sysroot is already passed by TOOLCHAIN_OPTIONS var which is appended to CC var already, so perhaps you need to adjust makefile to allow it use CC from env

File: recipes-bsp/boot0/boot0.bb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+EXTRA_OEMAKE = "CROSS_COMPILE=${TARGET_PREFIX} p=sun20iw1p1 mmc"
+EXTRA_OEMAKE += "STAGING_INCDIR=${STAGING_INCDIR_NATIVE}"
+
+GCC_INCLUDE = "${RECIPE_SYSROOT_NATIVE}${libdir}/${TARGET_SYS}/gcc/${TARGET_SYS}/*/include"
+
+inherit deploy
+
+do_configure () {
+	cp -r ${GCC_INCLUDE}/* ${STAGING_INCDIR_NATIVE}
```
Comment (kraj) [inline recipes-bsp/boot0/boot0.bb:27]: this either.

File: recipes-bsp/boot0/boot0.bb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+SRCREV = "0ad88bfdb723b1ac74cca96122918f885a4781ac"
+
+S = "${WORKDIR}/git"
+
+CFLAGS += " --sysroot=${RECIPE_SYSROOT}"
+
+EXTRA_OEMAKE = "CROSS_COMPILE=${TARGET_PREFIX} p=sun20iw1p1 mmc"
+EXTRA_OEMAKE += "STAGING_INCDIR=${STAGING_INCDIR_NATIVE}"
+
+GCC_INCLUDE = "${RECIPE_SYSROOT_NATIVE}${libdir}/${TARGET_SYS}/gcc/${TARGET_SYS}/*/include"
```
Comment (kraj) [inline recipes-bsp/boot0/boot0.bb:22]: this wont be needed.

File: recipes-bsp/opensbi/opensbi_%.bbappend [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Mainline OpenSBI supports the C906 out of the box, but it needs a few tweaks
+# and a new reset driver for the sunxi watchdog.
+FILESEXTRAPATHS:prepend:nezha := "${THISDIR}/files:"
+
+SRC_URI:append:nezha = " \
+    file://0001-lib-utils-fdt-Require-match-data-to-be-const.patch \
+    file://0002-lib-utils-timer-Add-a-separate-compatible-for-the-D1.patch \
+"
+
+EXTRA_OEMAKE:nezha = "PLATFORM=${RISCV_SBI_PLAT} I=${D} FW_PIC=y CLANG_TARGET= "
```
Comment (kraj) [inline recipes-bsp/opensbi/opensbi_%.bbappend:10]: drop this, since OE-core sets it to same.

File: recipes-bsp/opensbi/opensbi_%.bbappend [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+do_install:prepend:nezha () {
+    autotools_do_install
```
Comment (kraj) [inline recipes-bsp/opensbi/opensbi_%.bbappend:13]: why is this needed ?

File: recipes-kernel/linux/linux-nezha-dev.bb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+           file://enable-cgroups.cfg \
+           file://enable-modules.cfg \
+          "
+
+LINUX_VERSION ?= "5.16.0"
+LINUX_VERSION_EXTENSION:append = "-nezha"
+
+KBUILD_DEFCONFIG = "nezha_defconfig"
+
+COMPATIBLE_MACHINE = "(nezha-allwinner-d1)"
```
Comment (kraj) [inline recipes-kernel/linux/linux-nezha-dev.bb:22]: there is no setting for KERNEL_DEVICETREE variable here. Is it not needed ?

File: recipes-kernel/linux/linux-nezha-dev.bb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+require recipes-kernel/linux/linux-mainline-common.inc
+FILESEXTRAPATHS:prepend = "${FILE_DIRNAME}/linux-nezha:"
+SUMMARY = "Nezha dev kernel recipe"
+
+KERNEL_VERSION_SANITY_SKIP = "1"
```
Comment (kraj) [inline recipes-kernel/linux/linux-nezha-dev.bb:5]: since you are pinning SRCREV to a valid SHA its not needed.

File: recipes-kernel/linux/linux-nezha-dev.bb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+SRCREV = "af3f4a1caec12845b809fba959e6334ab3b52a40"
+FORK ?= "tekkamanninja"
+BRANCH ?= "allwinner_nezha_d1_devel"
+SRC_URI = "git://github.com/${FORK}/linux.git;protocol=https;branch=${BRANCH} \
+           file://0001-riscv-fix-build-with-binutils-2.38.patch \
+           file://enable-autofs4.cfg \
+           file://enable-cgroups.cfg \
+           file://enable-modules.cfg \
```
Comment (kraj) [inline recipes-kernel/linux/linux-nezha-dev.bb:14]: instead of cfg files can we use the KERNEL_FEATURES variable as much as possible ? see
https://git.yoctoproject.org/yocto-kernel-cache/tree/
for possible options.

File: recipes-kernel/linux/linux-nezha/enable-cgroups.cfg [hunk 10]
```diff
@@ -0,0 +1,10 @@
+CONFIG_CGROUPS=y
```
Comment (kraj) [inline recipes-kernel/linux/linux-nezha/enable-cgroups.cfg:1]: instead of this file you can use
KERNEL_FEATURES += "features/cgroups/cgroups.cfg"

**Review by kraj — COMMENTED** [review]
This looks very good and thanks a bunch for your contributions. These are comments that an be submitted as followup improvement too.

File: recipes-kernel/linux/linux-nezha-dev.bb (hunk 7 above)
Comment (smaeul) [reply recipes-kernel/linux/linux-nezha-dev.bb:22]: The devicetree is provided by platform firmware (u-boot). It is not loaded from disk.

File: conf/machine/nezha-allwinner-d1.conf [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+PREFERRED_VERSION_openocd-native = "riscv"
+PREFERRED_VERSION_openocd = "riscv"
+
+EXTRA_IMAGEDEPENDS += "opensbi"
+RISCV_SBI_PLAT = "generic"
+
+# Depends on which version of the Nezha board is available, you can choose
+# "512m", "1g" or "2g". It is necessary to point proper Device Tree Blob for
+# U-Boot, OpenSBI or Linux kernel.
+NEZHA_DRAM_SIZE ?= "1g"
```
Comment (smaeul) [inline conf/machine/nezha-allwinner-d1.conf:26]: This is an awful hack which will never be merged upstream, and since https://github.com/smaeul/sun20i_d1_spl/commit/525883d3721f4c4d78b498e780b44e85d0676abf it is entirely unnecessary. Please use the DTB that is already provided by U-Boot in RAM (at `$fdtcontroladdr`), and things will Just Work for every board variant.

File: recipes-bsp/boot0/boot0.bb (hunk 3 above)
Comment (Cezarus27) [reply recipes-bsp/boot0/boot0.bb:27]: This is necessary to avoid compilation errors. The `STAGING_INCDIR_NATIVE` doesn't have a `stdbool.h` file and others which I don't remember now:
```
| In file included from fdt_wip.c:6:
| libfdt_env.h:10:10: fatal error: stdbool.h: No such file or directory
|    10 | #include <stdbool.h>
|       |          ^~~~~~~~~~~
| compilation terminated.
```

This is my workaround for this problem.

File: conf/machine/nezha-allwinner-d1.conf (hunk 11 above)
Comment (Cezarus27) [reply conf/machine/nezha-allwinner-d1.conf:26]: It will be difficult to use `fitImage` in this case as it has a built-in DTB. Maybe then I should use `uImage` format and load the kernel with the DTB loaded from RAM?

File: recipes-bsp/boot0/boot0.bb (hunk 1 above)
Comment (Cezarus27) [reply recipes-bsp/boot0/boot0.bb]: I apply your suggestion.

File: recipes-bsp/opensbi/opensbi_%.bbappend (hunk 6 above)
Comment (Cezarus27) [reply recipes-bsp/opensbi/opensbi_%.bbappend]: I checked and it is not used so I removed it.

File: recipes-kernel/linux/linux-nezha-dev.bb (hunk 9 above)
Comment (Cezarus27) [reply recipes-kernel/linux/linux-nezha-dev.bb]: Okey, I didn't know about it earlier. It is replaced with this:
```
KERNEL_FEATURES += "features/cgroups/cgroups.cfg"
KERNEL_FEATURES += "ktypes/standard/standard.cfg"
```

File: recipes-bsp/boot0/boot0.bb (hunk 2 above)
Comment (Cezarus27) [reply recipes-bsp/boot0/boot0.bb]: I apply your suggestion with an updated patch:
```
recipes-bsp/boot0/files/0001-config.mk-provide-path-to-linux-headers.patch
```

**Review by alistair23 — CHANGES_REQUESTED** [review]

File: conf/machine/nezha-allwinner-d1.conf (hunk 11 above)
Comment (smaeul) [reply conf/machine/nezha-allwinner-d1.conf:26]: From what I understand, it should be possible to build and boot a FIT image that does not contain a DTB.

File: recipes-bsp/opensbi/files/0002-lib-utils-timer-Add-a-separate-compatible-for-the-D1.patch [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+Date: Sun, 8 Aug 2021 01:18:44 -0500
+Subject: [PATCH 2/2] lib: utils/timer: Add a separate compatible for the D1
+ CLINT
+
+The CLINT in the Allwinner D1 SoC apparently does not support 64-bit
+MMIO access. A property was added to support this quirk (and that
+property was copied to the ACLINT MTIMER code). However, since this
+difference in behavior makes the D1 CLINT incompatible with the SiFive
+CLINT's programming interface, a better solution is to use a separate
+compatible string.
```
[the remark this answers is not in the document]
Comment (Cezarus27) [reply recipes-bsp/opensbi/files/0002-lib-utils-timer-Add-a-separate-compatible-for-the-D1.patch:12]: Maybe @smaeul can explain that better than me. What I can do is to check and confirm that without this change board will boot.

File: recipes-bsp/opensbi/files/0001-lib-utils-fdt-Require-match-data-to-be-const.patch [hunk 13]
```diff
@@ -0,0 +1,66 @@
+From 65410b015fc16d1c47b1972f89a35f5c8dd10abe Mon Sep 17 00:00:00 2001
+From: Samuel Holland <samuel@sholland.org>
+Date: Tue, 19 Oct 2021 19:25:37 -0500
+Subject: [PATCH 1/2] lib: utils/fdt: Require match data to be const
+
+Match data stores hardware attributes which do not change at runtime, so
+it does not need to be mutable. Make it const.
```
[the remark this answers is not in the document]
Comment (Cezarus27) [reply recipes-bsp/opensbi/files/0001-lib-utils-fdt-Require-match-data-to-be-const.patch:7]: Same as earlier. @smaeul

File: recipes-bsp/opensbi/files/0002-lib-utils-timer-Add-a-separate-compatible-for-the-D1.patch (hunk 12 above)
[the remark this answers is not in the document]
Comment (smaeul) [reply recipes-bsp/opensbi/files/0002-lib-utils-timer-Add-a-separate-compatible-for-the-D1.patch:12]: Mainline doesn't have a solution. The existing property was added to OpenSBI without any review of the binding, which was subsequently NAKed by the DT maintainers.

The current state is waiting for Anup to send a new version of the ACLINT MTIMER binding upstream, so the T-HEAD/D1 compatible can be added to it:

https://lore.kernel.org/lkml/20211007123632.697666-9-anup.patel@wdc.com/
https://lore.kernel.org/lkml/CAAhSdy1sTuEbjGM+XLF_Vg6=e91E31Yvh7js-h4c+Ru89DQJsw@mail.gmail.com/

So this patch implements the binding that will eventually be upstream. The other patch is a dependency of this one.

File: recipes-bsp/boot0/boot0.bb (hunk 3 above)
Comment (kraj) [reply recipes-bsp/boot0/boot0.bb]: Please apply patches from these pulls https://github.com/smaeul/sun20i_d1_spl/pull/8 and https://github.com/smaeul/sun20i_d1_spl/pull/7, after these patches your recipe should simplify dramatically, and all you would need perhaps is edit LDFLAGS  something like `EXTRA_OEMAKE += 'LDFLAGS='` would do it.

File: recipes-bsp/opensbi/files/0002-lib-utils-timer-Add-a-separate-compatible-for-the-D1.patch (hunk 12 above)
[the remark this answers is not in the document]
Comment (kraj) [reply recipes-bsp/opensbi/files/0002-lib-utils-timer-Add-a-separate-compatible-for-the-D1.patch:12]: OK, that would be fine. I think we can track its progress upstream.

**Comment by kraj** [comment]
@alistair23 are you fine with the replies ? I would like to merge this and perhaps send some improvements e.g. boot0 recipe would need a rewrite etc. but I would not like to hold it for that.

**Comment by Cezarus27** [comment]
@kraj @alistair23 @smaeul I adhered to your comments and applied the changes you point to. I think the recipes now are cleaner and simpler. I checked that the build passed without errors and OS is booting.

**Comment by kraj** [comment]
> @kraj @alistair23 @smaeul I adhered to your comments and applied the changes you point to. I think the recipes now are cleaner and simpler. I checked that the build passed without errors and OS is booting.

Can you update boot0 to latest SRCREV ? @smaeul has applied few of fixes I have sent could be helpful in simplifying the recipe.

**Review by alistair23 — APPROVED** [review]

File: conf/machine/nezha-allwinner-d1.conf [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+EXTRA_IMAGEDEPENDS += "opensbi"
+RISCV_SBI_PLAT = "generic"
+
+## This sets u-boot as the default OpenSBI payload
+### Nezha board uses TOC1 images loaded by SPL and containing U-Boot, DTB, and
+### OpenSBI so 'RISCV_SBI_PAYLOAD' isn't used.
+### 'RISCV_SBI_FDT' isn't used because the DTB is loaded from RAM at
+### ${fdtcontroladdr}
+#RISCV_SBI_PAYLOAD ?= "u-boot.bin"
+#RISCV_SBI_FDT ?= "sun20i-d1-nezha.dtb"
```
Comment (alistair23) [inline conf/machine/nezha-allwinner-d1.conf:29]: Why keep these around if they are unused?

**Comment by alistair23** [comment]
Looks good to me, happy to have it merged now

**Review by kraj — APPROVED** [review]
lgtm
